File: Week_06/java/src/_392_IsSubsequence.java
/**
 * 第一遍：2020/07/27周一 ✅
 * 第二遍：2020/07/26周日
 * 第三遍：2020/07/27周一
 * 第三遍：2020/06/21周日
 * 第四遍：2020/07/05周日
 */
class _392_IsSubsequence {
    public boolean isSubsequence(String s, String t) {
        if (s == null || t == null) return false;
        int m = s.length();
        int n = t.length();
        int i = 0, j = 0;
        while (i < m && j < n) {
            if (t.charAt(j) == s.charAt(i)) {
                i++;
            }
            j++;
        }
        return i == m;
    }

    class SolutionDp {
        /*
        如果有大量输入的 S，称作S1, S2, ... , Sk 其中 k >= 10亿，你需要依次检查它们是否为 T 的子序列。
        在这种情况下，你会怎样改变代码？
         */
        public boolean isSubsequence(String s, String t) {
            //f(i,j)表示从i开始往后字符j出现的位置
            //f(i,j) = i 如果i处的字符和j相等
            //i in [0, m] f(m, j) = m  j in [0, 25]
            if (s == null || t == null) return false;
            int m = t.length();
            int n = s.length();
            int[][] dp = new int[m + 1][26];
            for (int j = 0; j < 26; j++) {
                dp[m][j] = m;
            }
            for (int i = m - 1; i >= 0; i--) {
                for (int j = 0; j < 26; j++) {
                    if (t.charAt(i) == j + 'a') {
                        dp[i][j] = i;
                    } else {
                        dp[i][j] = dp[i + 1][j];
                    }
                }
            }
            int i = 0;
            for (int k = 0; k < n; k++) {
                int idx = dp[i][s.charAt(k) - 'a'];
                if (idx == m) {
                    return false;
                }
                i = idx + 1;
            }
            return true;
        }
    }
}

File: Week_06/kotlin/src/_392_IsSubsequence.kt
fun isSubsequence(s: String, t: String): Boolean {
    var i = 0
    var j = 0
    val m = s.length
    val n = t.length
    while (i < n && j < m) {
        if (t[i] == s[j]) {
            j++;
        }
        i++;
    }
    return j == m
}

fun isSubsequenceWithPreprocessing(s: String, t: String): Boolean {
    //我们用了大量的时间在t中查找s中某个字符出现的位置，那么能不能加速这一过程
    //也就是在t中每一个位置存储每一个字符下一次在t中出现的位置，也就是建立索引。
    //f(i,j)表示t中从i开始包含i往后字符j第一次出现的位置
    //如果j == t[i] f(i,j) = i 否则 f(i, j) = f(i + 1,j)
    //i in [0, t.length - 1] 边界是f(t.length - 1, j) = t.length
    val n = t.length
    val dp = Array(n + 1) { IntArray(26) }
    for (j in 0..25) {
        dp[n][j] = n
    }
    for (i in n - 1 downTo 0) {
        for (j in 0..25) {
            if (t[i] - 'a' == j) {
                dp[i][j] = i
            } else {
                dp[i][j] = dp[i + 1][j]
            }
        }
    }

    var i = 0
    for (c in s) {
        val idx = dp[i][c - 'a']
        if (idx == n) {
            return false
        }
        i = idx + 1
    }
    return true
}